### PR TITLE
MAINT: modernization of qhull.pyx

### PR DESCRIPTION
**Pull request by tylerjereddy**
The [devdocs detailed roadmap](https://scipy.github.io/devdocs/roadmap-detailed.html#scipy-roadmap-detailed) contains a mandate for code modernization:

> Cython’s old syntax for using NumPy arrays should be removed and replaced with Cython memoryviews.

This PR aims to do this in full for the Cython interface to Qhull. While I found no evidence of performance gains (see below) in a local test, there may be opportunities to release the GIL in more places now--the full complement of advantages over the old buffer syntax [are highlighted in the docs](https://cython.readthedocs.io/en/latest/src/userguide/memoryviews.html#comparison-to-the-old-buffer-support).

For reviewers to consider:
- this assumes that our tests are probing all attributes that are exposed from Qhull through to Python; any such attributes that aren't tested are prone to be returned as memoryviews instead of ndarrays proper; in-library usage of Qhull by some parts of `interpolate` exposed a few cases not covered by `spatial` tests but there are bound to be others
- ways to avoid `np.asarray` and `np.array` calls are probably useful -- I think some may be hard to avoid, but maybe I have a few more than needed
- some of my typing / casting changes may be too aggressive re: np.type vs type (but the tests don't complain in my hands)

<details>

`asv continuous -e -b "time_convex_hull" 983c6d >& bench.txt`

```
       before           after         ratio
     [f4f3c0ed]       [983c6d5f]
     <master>         <modernize_qhull_pyx>
+      69.0±0.8μs       84.5±0.7μs     1.22  spatial.ConvexHullBench.time_convex_hull(10, False)
+        74.0±1μs         90.3±1μs     1.22  spatial.ConvexHullBench.time_convex_hull(10, True)
+         155±3μs        172±0.9μs     1.11  spatial.ConvexHullBench.time_convex_hull(100, False)
+         163±1μs          181±3μs     1.11  spatial.ConvexHullBench.time_convex_hull(100, True)
```

</details>

### CODE REVIEW

**Comment by tylerjereddy** [comment]
I can see refguide is finding some untested code paths :)

**Comment by WarrenWeckesser** [comment]
The changes make the code 10 to 20 percent slower.  What is gained by making this change?  If it is just cosmetic (i.e. the newer syntax looks nicer), it doesn't seem worth it.

**Comment by perimosocordiae** [comment]
It's possible we're leaving some performance on the table by not specifying contiguous-access dimensions.

That is, we have `cdef double[:, :]` but might want to use `cdef double[:, ::1]` instead.

**Comment by larsoner** [comment]
While doing this we need to be careful not to break existing code e.g. with `joblib` parallelization, see #9993

**Comment by tylerjereddy** [comment]
I'll put a "needs-work" label on this--hopefully I can get it up to a standard that satisfies reviewers eventually. It was tricky just doing the initial surgery / getting tests passing.

**Comment by WarrenWeckesser** [comment]
>  It was tricky just doing the initial surgery / getting tests passing.

And thanks for doing that!

I did some elementary tests, and in the simplest cases (e.g. simply summing a contiguous 1-d read-only array), the `ndarray` and `memoryview` versions had the same performance.  The relative performance in more complicated examples can depend on how error-checking is configured,  what kind of slicing is done on the arrays, and whether or not new arrays are created.  And probably other things--I don't have a complete grasp of all the issues.  Jake van der Plas's blogs from 2012 on the issue are interesting reading:

* http://jakevdp.github.io/blog/2012/08/08/memoryview-benchmarks/
* http://jakevdp.github.io/blog/2012/08/16/memoryview-benchmarks-2/

**Comment by tylerjereddy** [comment]
One thing I've lost time on here is getting traction to line profile the Cython code. There are plenty of stack overflow questions / answers about this, but the machinery is informal at best I think using i.e., @rkern `line_profiler` and some extra directives.  Some nice examples are available in Juypter context, but I don't think a transplant of the baked-in .pyx to a notebook cell is suitable in this case.

In particular, I think I'd need to pass `CYTHON_TRACE=1` and `CYTHON_TRACE_NOGIL=1` to gcc via `define_macros`, but many of our Cython files go through our custom `cythonize.py`, which doesn't even use `define_macros` directly. There's probably a way around this, but it is not obvious and feels awkward to get going.

**Comment by tylerjereddy** [comment]
This didn't quite work out -- maybe needs to be done in smaller chunks, but probably not a priority so closing my own PR here for now.

File: scipy/spatial/qhull.pyx [hunk 1]
```diff
@@ -651,7 +651,7 @@ cdef class _Qhull:
 
                         if ncoplanar >= coplanar.shape[0]:
                             with gil:
-                                tmp = coplanar
+                                tmp = np.array(coplanar)
```
Comment (rainwoodman) [inline scipy/spatial/qhull.pyx:654]: This change adds a cast from memory view to ndarray  and a cast from ndarray to memoryview. which weren't there before. Could be contributing to the 10us slow down referred in #12379

**Pull request closed without merging**